File: python/sorting_algorithms/algos/quick_sort.py
from typing import Sequence, Optional
from ..base import _setup_tracer, _finalize, SortTracer, Number


def quick_sort(data: Sequence[Number], *, trace: bool = False, tracer: Optional[SortTracer] = None):
    arr = list(data)
    tracer, ret_trace = _setup_tracer("quick_sort", trace, tracer)

    def _partition(lo: int, hi: int) -> int:
        pivot = arr[hi]
        i = lo
        for j in range(lo, hi):
            if tracer:
                tracer.record(arr, (j, hi), "compare")
            if arr[j] <= pivot:
                arr[i], arr[j] = arr[j], arr[i]
                if tracer:
                    tracer.record(arr, (i, j), "swap")
                i += 1
        arr[i], arr[hi] = arr[hi], arr[i]
        if tracer:
            tracer.record(arr, (i,), "pivot set")
        return i

    def _quicksort(lo: int, hi: int):
        if lo < hi:
            p = _partition(lo, hi)
            _quicksort(lo, p - 1)
            _quicksort(p + 1, hi)

    _quicksort(0, len(arr) - 1)
    return _finalize(arr, tracer, ret_trace)

File: c/src/quick_sort.c
#include "../sorting_algorithms.h"

static size_t partition(int *arr, size_t lo, size_t hi) {
    int pivot = arr[hi];
    size_t i = lo;
    for (size_t j = lo; j < hi; ++j) {
        if (arr[j] <= pivot) {
            int tmp = arr[i]; arr[i] = arr[j]; arr[j] = tmp;
            ++i;
        }
    }
    int tmp = arr[i]; arr[i] = arr[hi]; arr[hi] = tmp;
    return i;
}

static void quick_sort_impl(int *arr, size_t lo, size_t hi) {
    if (lo < hi) {
        size_t p = partition(arr, lo, hi);
        if (p > 0) quick_sort_impl(arr, lo, p - 1);
        quick_sort_impl(arr, p + 1, hi);
    }
}

void quick_sort(int *arr, size_t n) {
    if (n > 0) quick_sort_impl(arr, 0, n - 1);
}
